### PR TITLE
WIP - Add organization admin

### PR DESCRIPTION
**Pull request by Vampouille**
This PR include work on server side and on UI side. It adds : 
* administration part for organization : delete, modify and create
* geometry part of organizations
* it change type parameter name : `orgType` to `type`
* validate fields on server side and UI side
* it rewrite fields validation class

### CODE REVIEW

**Comment by Vampouille** [comment]
Work in progress, please do not merge

File: ldapadmin/migrate-orgs.py [hunk 1]
```diff
@@ -88,7 +88,7 @@ def addUserToOrg(self, user_uid, org_cn):
     def createOrg(self, org_id, org_name):
         if not self.connection.add("cn=%s,%s" % (org_id, self.base_org_dn),
                                    ["groupOfMembers", "top"],
-                                   {"businessCategory" : "REGISTERED",
+                                   {"businessCategory" : "VALID",
```
Comment (Vampouille) [inline ldapadmin/migrate-orgs.py]: We need to unify values : choose between `VALID` and `REGISTERED`.

File: ldapadmin/src/main/java/org/georchestra/ldapadmin/dto/Org.java [hunk 2]
```diff
@@ -35,6 +35,9 @@
     public static final String JSON_STATUS = "status";
     public static final String JSON_MEMBERS = "members";
 
+    public static final String STATUS_REGISTERED = "REGISTERED";
```
Comment (Vampouille) [inline ldapadmin/src/main/java/org/georchestra/ldapadmin/dto/Org.java]: We can use this value : `REGISTERED`

**Comment by Vampouille** [comment]
@tonio it seems that 'required' attribute is not used when validating user form on admin part : `/ldapadmin/console/#/users/testuser/infos`. Form is posted even if email is empty.

**Comment by Vampouille** [comment]
> @tonio it seems that 'required' attribute is not used when validating user form on admin part : /ldapadmin/console/#/users/testuser/infos. Form is posted even if email is empty.

Should be fixed with : 5167f6b, @tonio can you review this part ?

**Comment by Vampouille** [comment]
@tonio Label of last entry of group droplist here : `/ldapadmin/console/#/analytics/all` is not displayed instead it display : `{{group.cn}}`. Last entry is 'TEMPORARY' group. Do you have an idea ?

**Comment by Vampouille** [comment]
@tonio Add just add translation for 'Tous' but I don't how to test it. Ui never display this string :

https://github.com/georchestra/georchestra/blob/master/ldapadmin/src/main/webapp/console/app/components/groups/groups.tpl.html#L32

Can we remove it ?

**Comment by Vampouille** [comment]
I try to use i18n files for this : 
https://github.com/georchestra/georchestra/blob/master/ldapadmin/src/main/webapp/console/app/components/logs/logs.es6#L12
With following code, but it does not work :
```js
    let i18n = {}
    this.$injector.get('translate')('logs.error',i18n)
    let msg = i18n.error
```
@tonio Do you known what I miss ?

**Comment by Vampouille** [comment]
Last item of recipient droplist on /ldapadmin/console/#/logs has label `{{::target}}` @tonio do you known how to fix this ?

**Comment by Vampouille** [comment]
We need to make configurable : 
* initial map position and zoom for orgs area
* how to group area
* ows resource to fetch area geometry
* field to use on ows resource to store area (ex: insee_code)

**Comment by Vampouille** [comment]
@tonio on /ldapadmin/account/new page angular translatino is not working and style :
```css
.area .ol-dragbox {  border-color: #009FE3; }
```
is not apply to selection rectangle. Can you check this ?

**Comment by Vampouille** [comment]
TODO : 

 - [ ]  Label of last entry of group droplist here : `/ldapadmin/console/#/analytics/all` is not displayed instead it display : `{{group.cn}}`. Last entry is 'TEMPORARY' group.
 - [ ] Last item of recipient droplist on /ldapadmin/console/#/logs has label {{::target}}
 - [ ] `.area .ol-dragbox {  border-color: #009FE3; }` is not apply to selection rectangle.
 - [ ]  We need to make configurable : 
  * initial map position and zoom for orgs area
  * how to group area
  * ows resource to fetch area geometry
  * field to use on ows resource to store area (ex: insee_code)

**Comment by Vampouille** [comment]
Work is finished, please review !

**Comment by Vampouille** [comment]
@fgravin @fvanderbiest Can you review this PR ?

File: ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/users/UsersController.java [hunk 3]
```diff
@@ -558,8 +560,8 @@ public void delete(@PathVariable String uid, HttpServletRequest request, HttpSer
     public void getUserRequiredFields(HttpServletResponse response) throws IOException{
         try {
             JSONArray fields = new JSONArray();
-            for(String field : this.validation.getUserRequiredFields())
-                fields.put(field);
+			fields.put("uid");
+			fields.put("mail");
```
Comment (fvanderbiest) [inline ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/users/UsersController.java]: indents ?

Comment (Vampouille) [inline ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/users/UsersController.java]: yes tab instead of spaces ... I will fix this...

Comment (Vampouille) [inline ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/users/UsersController.java]: Fix by https://github.com/georchestra/georchestra/pull/1560/commits/095ff952704ca7bca307cfa84d2970e63d496495

File: ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/log/LogController.java [hunk 4]
```diff
@@ -102,4 +106,15 @@ public String find( HttpServletRequest request, @PathVariable int limit, @PathVa
 
 		return new JSONObject().put("logs", res).toString(4);
 	}
+
+	@ExceptionHandler(Exception.class)
+	@ResponseStatus(value = HttpStatus.INTERNAL_SERVER_ERROR)
+	@ResponseBody
+	public String handleException(Exception e, HttpServletResponse response) throws JSONException {
+		LOG.error(e.getMessage());
+		response.setContentType("application/json; charset=utf-8");
```
Comment (fgravin) [inline ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/log/LogController.java]: minor: can't be done with `@RequestMapping` `produces` annotations ?

File: ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/orgs/OrgsController.java [hunk 5]
```diff
@@ -315,8 +338,7 @@ public void deleteOrg(@PathVariable String commonName, HttpServletResponse respo
     public void getUserRequiredFields(HttpServletResponse response) throws IOException{
         try {
```
Comment (fgravin) [inline ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/orgs/OrgsController.java]: Think of removing the `try catch` blocks if the `@ExceptionHandler` is used.

File: ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/orgs/OrgsController.java [hunk 6]
```diff
@@ -347,6 +369,58 @@ public void getOrgTypeVAlues(HttpServletResponse response) throws IOException{
...
+        areas.put("group", this.georConfig.getProperty("AreasGroup"));
+        res.put("areas", areas);
+        ResponseUtil.buildResponse(response, res.toString(4), HttpServletResponse.SC_OK);
+    }
+
+
+    @ExceptionHandler(Exception.class)
+    @ResponseStatus(value = HttpStatus.INTERNAL_SERVER_ERROR)
+    @ResponseBody
+    public String handleException(Exception e, HttpServletResponse response) throws JSONException, IOException {
```
Comment (fgravin) [inline ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/orgs/OrgsController.java]: For me, if you use `@responseBody` in your handler methods, a thrown exception won't return a 200 code status.

File: ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/users/UsersController.java [hunk 7]
```diff
@@ -691,59 +693,50 @@ private Account modifyAccount(Account account, ServletInputStream inputStream) t
...
+		String street        = RequestUtil.getFieldValue(json, UserSchema.STREET_KEY);
+		String locality      = RequestUtil.getFieldValue(json, UserSchema.LOCALITY_KEY);
+		String phone         = RequestUtil.getFieldValue(json, UserSchema.TELEPHONE_KEY);
+		String facsimile     = RequestUtil.getFieldValue(json, UserSchema.FACSIMILE_KEY);
+		String title         = RequestUtil.getFieldValue(json, UserSchema.TITLE_KEY);
+		String description   = RequestUtil.getFieldValue(json, UserSchema.DESCRIPTION_KEY);
+		String manager       = RequestUtil.getFieldValue(json, UserSchema.MANAGER_KEY);
+		String org           = RequestUtil.getFieldValue(json, UserSchema.ORG_KEY);
+
+		if(givenName == null)
```
Comment (fgravin) [inline ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/users/UsersController.java]: _Question_:
why not testing the `json` against `Validation` ?

File: ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/users/UsersController.java [hunk 8]
```diff
@@ -691,59 +693,50 @@ private Account modifyAccount(Account account, ServletInputStream inputStream) t
 	 */
 	private Account createAccountFromRequestBody(ServletInputStream is) throws IllegalArgumentException, IOException {
```
Comment (fgravin) [inline ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/users/UsersController.java]: Why create account in user controller ?
Nothing redondant with `create` method of `NewAccountFormController` ?

File: ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/newaccount/NewAccountFormController.java [hunk 9]
```diff
@@ -153,20 +161,52 @@ public String create(HttpServletRequest request,
...
+			// A valid user identifier (uid) can only contain characters, numbers, hyphens or dot.
+			// It must begin with a character.
+			Pattern regexp = Pattern.compile("[a-zA-Z][a-zA-Z0-9.-]*");
+			Matcher m = regexp.matcher(formBean.getUid());
+			if(!m.matches())
+				result.rejectValue("uid", "uid.error.invalid", "required");
+		}
+
+		// first name and surname validation
+		if(!this.validation.validateUserField("firstName", formBean.getFirstName()))
```
Comment (fgravin) [inline ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/newaccount/NewAccountFormController.java]: Can't factorize this, or put this in `Validation` ?

File: ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/log/LogController.java (hunk 4 above)
Comment (Vampouille) [inline ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/log/LogController.java]: I don't known if annotation of original will be used.

File: ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/users/UsersController.java (hunk 7 above)
Comment (Vampouille) [inline ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/users/UsersController.java]: Validation is only used for end user input, admin users (MOD_LDAPADMIN) should be able to edit user without restriction.

**Pull request merged**